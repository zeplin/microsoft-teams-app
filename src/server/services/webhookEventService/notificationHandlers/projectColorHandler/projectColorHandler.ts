import {
    WebhookEvent,
    ProjectColorCreateEvent,
    ProjectColorUpdateEvent
} from "../../../../adapters/zeplin/types";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../config";
import { URL } from "url";
import { getMacAppRedirectURL } from "../getMacAppRedirectURL";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";

type Event = ProjectColorCreateEvent | ProjectColorUpdateEvent;

class ProjectColorHandler extends NotificationHandler<Event> {
    delay = SHORT_DELAY;
    private getText(events: Event[]): string {
        const [{
            payload: {
                action,
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        name: pivotColorName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added" : "updated";
        return events.length === 1
            ? md`**${pivotColorName}** is ${actionText} in _${projectName}_! ${getRandomEmoji()}`
            : md`**${events.length} new colors** are ${actionText} in _${projectName}_! ${getRandomEmoji()}`;
    }

    private getWebappURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/styleguide/colors`;
        events.forEach(event => webappURL.searchParams.append("cid", event.payload.resource.id));
        return webappURL.toString();
    }

    private getMacAppURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        return getMacAppRedirectURL(`${ZEPLIN_MAC_APP_URL_SCHEME}://colors?pid=${projectId}&cids=${events.map(event => event.payload.resource.id).join(",")}`);
    }

    shouldHandleEvent(event: WebhookEvent): event is Event {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: Event[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Make sure your stylesheets are up to date!"
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const projectColorHandler = new ProjectColorHandler();
