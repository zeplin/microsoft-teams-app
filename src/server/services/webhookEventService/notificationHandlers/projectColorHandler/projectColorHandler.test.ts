import {
    Project,
    ProjectColorCreatedEvent,
    ProjectColorUpdatedEvent
} from "@zeplin/sdk";

import { projectColorHandler } from "./projectColorHandler";

type GetDummyEventParams = {
    action?: string;
    colorId?: string;
    colorName?: string;
    projectPlatform?: Project["platform"];
}

function getDummyEvent({
    action = "created",
    colorId = "colorId",
    colorName = "colorName",
    projectPlatform = "web"
}: GetDummyEventParams = {}): ProjectColorCreatedEvent | ProjectColorUpdatedEvent {
    return {
        action,
        context: {
            project: {
                id: "projectId",
                name: "projectName",
                platform: projectPlatform
            }
        },
        resource: {
            id: colorId,
            data: {
                id: colorId,
                name: colorName
            }
        }
    } as ProjectColorCreatedEvent | ProjectColorUpdatedEvent;
}

describe("projectColorHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectColorHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectColorHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Project["platform"]>([
                "android",
                "ios",
                "macos",
                "web"
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectColorHandler.getTeamsMessage([getDummyEvent({ projectPlatform })])
                    ).toMatchSnapshot();
                }
            );
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectColorHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectColorHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each<Project["platform"]>([
                "android",
                "ios",
                "macos",
                "web"
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectColorHandler.getTeamsMessage([
                            getDummyEvent({
                                action: "updated",
                                projectPlatform
                            })
                        ])
                    ).toMatchSnapshot();
                }
            );
        });
    });
});
