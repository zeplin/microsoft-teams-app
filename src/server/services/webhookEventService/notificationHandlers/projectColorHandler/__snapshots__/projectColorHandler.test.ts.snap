// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectColorHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 new colors** are added in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fpid%3DprojectId%26cids%3DcolorId%2CanotherColorId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/colors?cid=colorId&cid=anotherColorId"
        }
      ]
    }
  ]
}
`;

exports[`projectColorHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is added in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fpid%3DprojectId%26cids%3DcolorId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/colors?cid=colorId"
        }
      ]
    }
  ]
}
`;

exports[`projectColorHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 new colors** are updated in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fpid%3DprojectId%26cids%3DcolorId%2CanotherColorId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/colors?cid=colorId&cid=anotherColorId"
        }
      ]
    }
  ]
}
`;

exports[`projectColorHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**colorName** is updated in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fcolors%3Fpid%3DprojectId%26cids%3DcolorId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/colors?cid=colorId"
        }
      ]
    }
  ]
}
`;
